File: src/lib/stores/projectType.ts
import { derived, writable } from 'svelte/store';

export type ProjectType =
	| 'software'
	| 'marketing'
	| 'construction'
	| 'journalism'
	| 'electrical'
	| 'business'
	| 'general';

export type ProjectTypeConfig = {
	type: ProjectType;
	displayName: string;
	groupTerm: string;
	groupTermPlural: string;
	taskTerm: string;
	taskTermPlural: string;
	memberTerm: string;
	statusOptions: string[];
	defaultRoles: string[];
	description: string;
	icon: string;
};

export type CurrentWorkspace = {
	id: string;
	name?: string;
	project_type?: string;
};

export const PROJECT_TYPE_CONFIGS: Record<ProjectType, ProjectTypeConfig> = {
	software: {
		type: 'software',
		displayName: 'Software Development',
		groupTerm: 'Sprint',
		groupTermPlural: 'Sprints',
		taskTerm: 'Task',
		taskTermPlural: 'Tasks',
		memberTerm: 'Member',
		statusOptions: ['Backlog', 'Todo', 'In Progress', 'In Review', 'Done'],
		defaultRoles: [
			'Backend Developer',
			'Frontend Developer',
			'DevOps Engineer',
			'QA Engineer',
			'UI/UX Designer',
			'Product Manager'
		],
		description: 'Tasks grouped into Sprints',
		icon: '💻'
	},
	marketing: {
		type: 'marketing',
		displayName: 'Marketing',
		groupTerm: 'Campaign',
		groupTermPlural: 'Campaigns',
		taskTerm: 'Deliverable',
		taskTermPlural: 'Deliverables',
		memberTerm: 'Member',
		statusOptions: ['Planned', 'In Progress', 'In Review', 'Published', 'Done'],
		defaultRoles: [
			'Campaign Manager',
			'Copywriter',
			'Designer',
			'SEO Specialist',
			'Social Media Manager',
			'Analyst'
		],
		description: 'Deliverables grouped into Campaigns',
		icon: '📣'
	},
	construction: {
		type: 'construction',
		displayName: 'Construction',
		groupTerm: 'Phase',
		groupTermPlural: 'Phases',
		taskTerm: 'Work Package',
		taskTermPlural: 'Work Packages',
		memberTerm: 'Crew Member',
		statusOptions: ['Planned', 'Mobilising', 'In Progress', 'Inspection', 'Complete'],
		defaultRoles: [
			'Project Manager',
			'Site Engineer',
			'Architect',
			'Safety Officer',
			'Electrician',
			'Structural Engineer'
		],
		description: 'Work Packages grouped into Phases',
		icon: '🏗️'
	},
	journalism: {
		type: 'journalism',
		displayName: 'Journalism',
		groupTerm: 'Assignment',
		groupTermPlural: 'Assignments',
		taskTerm: 'Story',
		taskTermPlural: 'Stories',
		memberTerm: 'Reporter',
		statusOptions: ['Pitch', 'Researching', 'Writing', 'Editing', 'Published'],
		defaultRoles: ['Editor', 'Reporter', 'Photographer', 'Fact-checker', 'Copy Editor', 'Producer'],
		description: 'Stories grouped into Assignments',
		icon: '📰'
	},
	electrical: {
		type: 'electrical',
		displayName: 'Electrical Engineering',
		groupTerm: 'Stage',
		groupTermPlural: 'Stages',
		taskTerm: 'Work Order',
		taskTermPlural: 'Work Orders',
		memberTerm: 'Technician',
		statusOptions: ['Planned', 'Scheduled', 'In Progress', 'Testing', 'Commissioned'],
		defaultRoles: [
			'Electrical Engineer',
			'Technician',
			'Inspector',
			'Project Manager',
			'CAD Designer',
			'Safety Officer'
		],
		description: 'Work Orders grouped into Stages',
		icon: '⚡'
	},
	business: {
		type: 'business',
		displayName: 'Business / Operations',
		groupTerm: 'Initiative',
		groupTermPlural: 'Initiatives',
		taskTerm: 'Action Item',
		taskTermPlural: 'Action Items',
		memberTerm: 'Member',
		statusOptions: ['Proposed', 'Approved', 'In Progress', 'On Hold', 'Closed'],
		defaultRoles: [
			'Operations Manager',
			'Business Analyst',
			'Stakeholder',
			'Finance Lead',
			'HR Lead',
			'Executive Sponsor'
		],
		description: 'Action Items grouped into Initiatives',
		icon: '📊'
	},
	general: {
		type: 'general',
		displayName: 'General',
		groupTerm: 'Phase',
		groupTermPlural: 'Phases',
		taskTerm: 'Task',
		taskTermPlural: 'Tasks',
		memberTerm: 'Member',
		statusOptions: ['Todo', 'In Progress', 'Done'],
		defaultRoles: ['Lead', 'Member', 'Contributor'],
		description: 'Tasks grouped into Phases',
		icon: '📋'
	}
};

export const currentWorkspace = writable<CurrentWorkspace | null>(null);

function normalizeProjectType(value: string | undefined | null): ProjectType {
	const normalized = (value ?? '').trim().toLowerCase() as ProjectType;
	return normalized in PROJECT_TYPE_CONFIGS ? normalized : 'general';
}

export const projectTypeConfig = derived(currentWorkspace, ($workspace) => {
	return PROJECT_TYPE_CONFIGS[normalizeProjectType($workspace?.project_type)];
});

File: backend/internal/models/project_type.go
package models

import (
	"strings"
	"time"
)

const DefaultProjectType = "software"

type ProjectTypeConfig struct {
	Type            string
	DisplayName     string
	GroupTerm       string
	GroupTermPlural string
	TaskTerm        string
	TaskTermPlural  string
	MemberTerm      string
	StatusOptions   []string
	DefaultRoles    []string
	Description     string
}

type Group struct {
	WorkspaceID  string    `json:"workspace_id"`
	GroupID      string    `json:"group_id"`
	Name         string    `json:"name"`
	DisplayOrder int       `json:"display_order"`
	StartDate    string    `json:"start_date"`
	EndDate      string    `json:"end_date"`
	Description  string    `json:"description"`
	CreatedAt    time.Time `json:"created_at"`
}

var ProjectTypeConfigs = map[string]ProjectTypeConfig{
	"software": {
		Type:            "software",
		DisplayName:     "Software Development",
		GroupTerm:       "Sprint",
		GroupTermPlural: "Sprints",
		TaskTerm:        "Task",
		TaskTermPlural:  "Tasks",
		MemberTerm:      "Member",
		StatusOptions:   []string{"Backlog", "Todo", "In Progress", "In Review", "Done"},
		DefaultRoles:    []string{"Backend Developer", "Frontend Developer", "DevOps Engineer", "QA Engineer", "UI/UX Designer", "Product Manager"},
		Description:     "Tasks grouped into Sprints",
	},
	"marketing": {
		Type:            "marketing",
		DisplayName:     "Marketing",
		GroupTerm:       "Campaign",
		GroupTermPlural: "Campaigns",
		TaskTerm:        "Deliverable",
		TaskTermPlural:  "Deliverables",
		MemberTerm:      "Member",
		StatusOptions:   []string{"Planned", "In Progress", "In Review", "Published", "Done"},
		DefaultRoles:    []string{"Campaign Manager", "Copywriter", "Designer", "SEO Specialist", "Social Media Manager", "Analyst"},
		Description:     "Deliverables grouped into Campaigns",
	},
	"construction": {
		Type:            "construction",
		DisplayName:     "Construction",
		GroupTerm:       "Phase",
		GroupTermPlural: "Phases",
		TaskTerm:        "Work Package",
		TaskTermPlural:  "Work Packages",
		MemberTerm:      "Crew Member",
		StatusOptions:   []string{"Planned", "Mobilising", "In Progress", "Inspection", "Complete"},
		DefaultRoles:    []string{"Project Manager", "Site Engineer", "Architect", "Safety Officer", "Electrician", "Structural Engineer"},
		Description:     "Work Packages grouped into Phases",
	},
	"journalism": {
		Type:            "journalism",
		DisplayName:     "Journalism",
		GroupTerm:       "Assignment",
		GroupTermPlural: "Assignments",
		TaskTerm:        "Story",
		TaskTermPlural:  "Stories",
		MemberTerm:      "Reporter",
		StatusOptions:   []string{"Pitch", "Researching", "Writing", "Editing", "Published"},
		DefaultRoles:    []string{"Editor", "Reporter", "Photographer", "Fact-checker", "Copy Editor", "Producer"},
		Description:     "Stories grouped into Assignments",
	},
	"electrical": {
		Type:            "electrical",
		DisplayName:     "Electrical Engineering",
		GroupTerm:       "Stage",
		GroupTermPlural: "Stages",
		TaskTerm:        "Work Order",
		TaskTermPlural:  "Work Orders",
		MemberTerm:      "Technician",
		StatusOptions:   []string{"Planned", "Scheduled", "In Progress", "Testing", "Commissioned"},
		DefaultRoles:    []string{"Electrical Engineer", "Technician", "Inspector", "Project Manager", "CAD Designer", "Safety Officer"},
		Description:     "Work Orders grouped into Stages",
	},
	"business": {
		Type:            "business",
		DisplayName:     "Business / Operations",
		GroupTerm:       "Initiative",
		GroupTermPlural: "Initiatives",
		TaskTerm:        "Action Item",
		TaskTermPlural:  "Action Items",
		MemberTerm:      "Member",
		StatusOptions:   []string{"Proposed", "Approved", "In Progress", "On Hold", "Closed"},
		DefaultRoles:    []string{"Operations Manager", "Business Analyst", "Stakeholder", "Finance Lead", "HR Lead", "Executive Sponsor"},
		Description:     "Action Items grouped into Initiatives",
	},
	"general": {
		Type:            "general",
		DisplayName:     "General",
		GroupTerm:       "Phase",
		GroupTermPlural: "Phases",
		TaskTerm:        "Task",
		TaskTermPlural:  "Tasks",
		MemberTerm:      "Member",
		StatusOptions:   []string{"Todo", "In Progress", "Done"},
		DefaultRoles:    []string{"Lead", "Member", "Contributor"},
		Description:     "Tasks grouped into Phases",
	},
}

func GetProjectTypeConfig(projectType string) ProjectTypeConfig {
	if cfg, ok := ProjectTypeConfigs[strings.ToLower(strings.TrimSpace(projectType))]; ok {
		return cfg
	}
	return ProjectTypeConfigs["general"]
}

func NormalizeProjectType(projectType string) string {
	normalized := strings.ToLower(strings.TrimSpace(projectType))
	if normalized == "" {
		return DefaultProjectType
	}
	if _, ok := ProjectTypeConfigs[normalized]; ok {
		return normalized
	}
	return DefaultProjectType
}
